### PR TITLE
[heft-jest-plugin] Add logHeapUsage

### PR DESCRIPTION
**Pull request by woss**
## Summary

This PR brings a new flag `--log-heap-usage`, and a corresponding change to the Reporter to show the heap usage for each test.

Jest logHeapUsage documentation is [here](https://jestjs.io/docs/cli#--logheapusage)

## Details

Useful to debug memory leaks. 

## How it was tested

Manual testing.

## Impacted documentation

Comments in the code and `rush change` json are provided.

### CODE REVIEW

**Comment by woss** [comment]
@microsoft-github-policy-service agree

File: heft-plugins/heft-jest-plugin/src/HeftJestReporter.ts [hunk 1]
```diff
@@ -63,9 +63,15 @@ export default class HeftJestReporter implements Reporter {
...
+
+    // calculate memoryUsage to MB reference -> https://jestjs.io/docs/cli#--logheapusage
+    const memUsage: string = testResult.memoryUsage
+      ? ` ${Math.floor(testResult.memoryUsage / 1000000)}MB heap size`
+      : '';
+
     const message: string =
       ` ${this._getTestPath(test.path)} ` +
-      `(duration: ${duration}, ${numPassingTests} passed, ${numFailingTests} failed)`;
+      `(duration: ${duration}, ${numPassingTests} passed, ${numFailingTests} failed,${memUsage})`;
```
Comment (iclanton) [inline heft-plugins/heft-jest-plugin/src/HeftJestReporter.ts:74]: Put the comma in the conditional string too.

File: heft-plugins/heft-jest-plugin/heft-plugin.json [hunk 2]
```diff
@@ -36,6 +36,11 @@
           "argumentName": "SOURCE_FILE",
           "description": "Find and run the tests that cover a source file that was passed in as an argument.  This corresponds to the \"--findRelatedTests\" parameter in Jest's documentation. This parameter is not compatible with watch mode."
         },
+        {
+          "longName": "--log-heap-usage",
+          "parameterKind": "flag",
+          "description": "Logs the heap usage after every test. Useful to debug memory leaks. Use together with --expose-gc in node."
```
Comment (iclanton) [inline heft-plugins/heft-jest-plugin/heft-plugin.json:42]: Is this functionality useful if the `--expose-gc` parameter isn't provided? If it isn't, we should ensure it was passed and print a warning if it wasn't.

File: common/changes/@rushstack/heft-jest-plugin/add-jest-log-heap-usage_2023-09-12-12-20.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/heft-jest-plugin",
+      "comment": "add log-heap-usage flag",
```
Comment (iclanton) [inline common/changes/@rushstack/heft-jest-plugin/add-jest-log-heap-usage_2023-09-12-12-20.json:5]:
```suggestion
      "comment": "Add a `--log-heap-usage` flag that includes memory usage analysis in each test run.",
```

File: common/changes/@rushstack/heft-jest-plugin/add-jest-log-heap-usage_2023-09-12-12-20.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+      "type": "patch"
```
Comment (iclanton) [inline common/changes/@rushstack/heft-jest-plugin/add-jest-log-heap-usage_2023-09-12-12-20.json:6]:
```suggestion
      "type": "minor"
```

**Review by iclanton — APPROVED** [review]

File: heft-plugins/heft-jest-plugin/heft-plugin.json (hunk 2 above)
Comment (woss) [reply heft-plugins/heft-jest-plugin/heft-plugin.json:42]: I tested it without --expose-gc and it worked perfectly. I'm not really sure why they say this in Jest documentation when it's not absolutley neceseery. Mybe remove the mention of it for the time being.

File: heft-plugins/heft-jest-plugin/src/HeftJestReporter.ts (hunk 1 above)
Comment (woss) [reply heft-plugins/heft-jest-plugin/src/HeftJestReporter.ts]: @iclanton updated

File: heft-plugins/heft-jest-plugin/src/HeftJestReporter.ts [hunk 5]
```diff
@@ -172,7 +178,7 @@ export default class HeftJestReporter implements Reporter {
     // and print a newline.
     this._terminal.writeLine('\u001b[0m');
     this._terminal.writeLine(
-      `Run start. ${this._formatWithPlural(aggregatedResult.numTotalTestSuites, 'test suite', 'test suites')}`
+      `Run start. ${this._formatWithPlural(aggregatedResult.numTotalTestSuites, 'test suite', '68 suites')}`
```
Comment (iclanton) [inline heft-plugins/heft-jest-plugin/src/HeftJestReporter.ts:181]: 68?

File: heft-plugins/heft-jest-plugin/heft-plugin.json (hunk 2 above)
Comment (iclanton) [reply heft-plugins/heft-jest-plugin/heft-plugin.json:42]: Does that depend on the version of Node?

File: heft-plugins/heft-jest-plugin/src/HeftJestReporter.ts (hunk 5 above)
Comment (woss) [reply heft-plugins/heft-jest-plugin/src/HeftJestReporter.ts:181]: wow, i honestly have no idea what or how this ended up there.

**Pull request merged**